File: src/Kutochok/Models/Schema.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Kutochok.Models;

public enum Widget
{
    Text,
    TextArea,
    Markdown,
    Date,
    Number,
    List,
    Select,
    Url,
}

/// <summary>Як поле має виглядати у формі.</summary>
public sealed record FieldDef(
    string Name,
    string Label,
    Widget Widget,
    bool Required = false,
    string? Help = null,
    IReadOnlyList<SelectOption>? Options = null,
    double? Min = null,
    double? Max = null,
    double Step = 1);

public sealed record SelectOption(string Value, string Label);

/// <summary>
/// Markdown — один файл на запис у теці <see cref="Path"/>.
/// Yaml — спільний файл-список, ключ <see cref="RootKey"/>.
/// </summary>
public enum StorageKind
{
    Markdown,
    Yaml,
}

public sealed record CollectionDef(
    string Name,
    string Label,
    string Singular,
    StorageKind Kind,
    string Path,
    IReadOnlyList<FieldDef> Fields,
    string? RootKey = null,
    string? SortBy = null,
    string? Hint = null,
    /// <summary>Поле, за яким список ділиться на теки/групи із заголовками.</summary>
    string? GroupBy = null)
{
    public FieldDef? Body => Fields.FirstOrDefault(f => f.Widget == Widget.Markdown);

    public bool HasTags => Fields.Any(f => f.Name == "tags");
}

/// <summary>Опис усіх розділів застосунку.</summary>
public static class Schema
{
    private static readonly FieldDef Tags = new(
        "tags", "Tags", Widget.List,
        Help: "Comma separated. Used by the filter and search.");

    public static readonly IReadOnlyList<CollectionDef> Collections =
    [
        new CollectionDef(
            "notes", "Notes", "note", StorageKind.Markdown, "notes",
            SortBy: "date",
            Hint: "Longer writing: thoughts, drafts, journal",
            Fields:
            [
                new FieldDef("title", "Title", Widget.Text, Required: true),
                new FieldDef("date", "Date", Widget.Date, Required: true),
                Tags,
                new FieldDef("body", "Text", Widget.Markdown, Required: true),
            ]),

        new CollectionDef(
            "til", "TIL", "TIL", StorageKind.Markdown, "til",
            SortBy: "date",
            Hint: "Today I learned — small things worth keeping",
            Fields:
            [
                new FieldDef("title", "Title", Widget.Text, Required: true),
                new FieldDef("date", "Date", Widget.Date, Required: true),
                Tags,
                new FieldDef("source", "Source", Widget.Url, Help: "Where you learned it."),
                new FieldDef("body", "Text", Widget.Markdown, Required: true),
            ]),

        new CollectionDef(
            "books", "Books", "book", StorageKind.Yaml, "books.yaml",
            RootKey: "books",
            SortBy: "year",
            Hint: "Stats and the chart are computed automatically",
            Fields:
            [
                new FieldDef("title", "Title", Widget.Text, Required: true),
                new FieldDef("status", "Status", Widget.Select,
                    Help: "«Reading now» keeps the book at the top of the list.",
                    Options: [new SelectOption("read", "read"), new SelectOption("reading", "reading now")]),
                new FieldDef("year", "Year", Widget.Number, Min: 1900, Max: 2200, Step: 1),
                new FieldDef("rating", "Rating", Widget.Number, Min: 0, Max: 5, Step: 0.5),
                new FieldDef("note", "Thoughts", Widget.TextArea),
                new FieldDef("url", "Link", Widget.Url),
            ]),

        new CollectionDef(
            "projects", "Projects", "project", StorageKind.Yaml, "projects.yaml",
            RootKey: "projects",
            SortBy: "year",
            GroupBy: "folder",
            Hint: "Grouped into folders — type any name you like",
            Fields:
            [
                new FieldDef("title", "Title", Widget.Text, Required: true),
                new FieldDef("folder", "Folder", Widget.Text,
                    Help: "For example: Work, Side projects. Empty means «No folder»."),
                new FieldDef("description", "Description", Widget.TextArea, Required: true),
                new FieldDef("status", "Status", Widget.Select,
                    Options:
                    [
                        new SelectOption("active", "active"),
                        new SelectOption("done", "done"),
                        new SelectOption("paused", "paused"),
                    ]),
                new FieldDef("year", "Year", Widget.Number, Min: 1900, Max: 2200, Step: 1),
                new FieldDef("stack", "Stack", Widget.List, Help: "Comma separated: Unity, C#"),
                new FieldDef("url", "Link", Widget.Url),
                new FieldDef("repo", "Repository", Widget.Url),
            ]),

        new CollectionDef(
            "links", "Links", "link", StorageKind.Yaml, "links.yaml",
            RootKey: "links",
            GroupBy: "group",
            Hint: "Worth keeping and re-reading some day",
            Fields:
            [
                new FieldDef("title", "Title", Widget.Text, Required: true),
                new FieldDef("url", "Address", Widget.Url, Required: true),
                new FieldDef("group", "Group", Widget.Text, Help: "Blogs, Tools, …"),
                new FieldDef("note", "Note", Widget.TextArea),
            ]),
    ];

    public static CollectionDef? Find(string name) =>
        Collections.FirstOrDefault(c => string.Equals(c.Name, name, StringComparison.Ordinal));
}

File: src/admin/schema.ts
/**
 * Опис колекцій для адмінки.
 *
 * Це «людське» дзеркало src/content.config.ts: там описано, що дозволено
 * в даних, а тут — як воно має виглядати у формі. Якщо додаєш поле в
 * content.config.ts і хочеш редагувати його через /admin — додай і сюди.
 */

export type Widget =
  | 'text'
  | 'textarea'
  | 'markdown'
  | 'date'
  | 'number'
  | 'boolean'
  | 'list'
  | 'select'
  | 'url';

export interface Field {
  name: string;
  label: string;
  widget: Widget;
  required?: boolean;
  /** Тільки для widget: 'select' */
  options?: { value: string; label: string }[];
  /** Підказка під полем */
  help?: string;
  /** Тільки для widget: 'number' */
  min?: number;
  max?: number;
  step?: number;
}

export interface Collection {
  name: string;
  label: string;
  /** Однина — для кнопки «Новий …» */
  singular: string;
  /**
   * markdown — один файл на запис, тека в `path`
   * yaml     — один спільний файл-масив, шлях у `path`
   */
  kind: 'markdown' | 'yaml';
  path: string;
  /** Тільки для kind: 'yaml' — ключ, під яким у файлі лежить список записів */
  rootKey?: string;
  fields: Field[];
  /** Поле, за яким сортувати список (спадання) */
  sortBy?: string;
  /** Чи можна створювати й видаляти записи */
  canCreate: boolean;
  canDelete: boolean;
  /** Підпис угорі списку */
  hint?: string;
}

const TAGS: Field = {
  name: 'tags',
  label: 'Теги',
  widget: 'list',
  help: 'Через кому. Сторінки тегів створюються самі.',
};

export const COLLECTIONS: Collection[] = [
  {
    name: 'posts',
    label: 'Пости',
    singular: 'пост',
    kind: 'markdown',
    path: 'src/content/posts',
    sortBy: 'date',
    canCreate: true,
    canDelete: true,
    hint: 'Ім’я файлу стає адресою: nazva.md → /writing/nazva',
    fields: [
      { name: 'title', label: 'Заголовок', widget: 'text', required: true },
      { name: 'date', label: 'Дата', widget: 'date', required: true },
      {
        name: 'description',
        label: 'Опис',
        widget: 'textarea',
        help: 'Одне речення для головної, RSS і пошуку. Необов’язково.',
      },
      TAGS,
      {
        name: 'draft',
        label: 'Чернетка',
        widget: 'boolean',
        help: 'Чернетку видно тільки локально, у зібраний сайт вона не потрапляє.',
      },
      { name: 'body', label: 'Текст', widget: 'markdown', required: true },
    ],
  },
  {
    name: 'til',
    label: 'TIL',
    singular: 'замітку',
    kind: 'markdown',
    path: 'src/content/til',
    sortBy: 'date',
    canCreate: true,
    canDelete: true,
    hint: 'Короткі замітки. Показуються повністю прямо в списку на /til',
    fields: [
      { name: 'title', label: 'Заголовок', widget: 'text', required: true },
      { name: 'date', label: 'Дата', widget: 'date', required: true },
      TAGS,
      { name: 'source', label: 'Джерело', widget: 'url', help: 'Посилання, звідки дізнався.' },
      { name: 'body', label: 'Текст', widget: 'markdown', required: true },
    ],
  },
  {
    name: 'pages',
    label: 'Сторінки',
    singular: 'сторінку',
    kind: 'markdown',
    path: 'src/content/pages',
    canCreate: true,
    canDelete: true,
    hint: 'about.md → /about. Щоб сторінка з’явилася в меню, додай її в src/site.ts',
    fields: [
      { name: 'title', label: 'Заголовок', widget: 'text', required: true },
      { name: 'description', label: 'Опис', widget: 'textarea' },
      {
        name: 'updated',
        label: 'Оновлено',
        widget: 'date',
        help: 'Показується під заголовком. Можна лишити порожнім.',
      },
      { name: 'body', label: 'Текст', widget: 'markdown', required: true },
    ],
  },
  {
    name: 'books',
    label: 'Книги',
    singular: 'книжку',
    kind: 'yaml',
    path: 'src/data/books.yaml',
    rootKey: 'books',
    sortBy: 'year',
    canCreate: true,
    canDelete: true,
    hint: 'Статистика й графік на /books рахуються самі',
    fields: [
      { name: 'title', label: 'Назва', widget: 'text', required: true },
      {
        name: 'status',
        label: 'Статус',
        widget: 'select',
        options: [
          { value: 'read', label: 'прочитано' },
          { value: 'reading', label: 'читаю зараз' },
        ],
        help: '«Читаю зараз» показується у блоці на головній.',
      },
      { name: 'year', label: 'Рік', widget: 'number', min: 1900, max: 2200, step: 1 },
      { name: 'rating', label: 'Оцінка', widget: 'number', min: 0, max: 5, step: 0.5 },
      { name: 'note', label: 'Враження', widget: 'textarea' },
      { name: 'url', label: 'Посилання', widget: 'url' },
    ],
  },
  {
    name: 'projects',
    label: 'Проєкти',
    singular: 'проєкт',
    kind: 'yaml',
    path: 'src/data/projects.yaml',
    rootKey: 'projects',
    sortBy: 'year',
    canCreate: true,
    canDelete: true,
    fields: [
      { name: 'title', label: 'Назва', widget: 'text', required: true },
      { name: 'description', label: 'Опис', widget: 'textarea', required: true },
      {
        name: 'status',
        label: 'Статус',
        widget: 'select',
        options: [
          { value: 'active', label: 'в роботі' },
          { value: 'done', label: 'завершено' },
          { value: 'paused', label: 'на паузі' },
        ],
      },
      { name: 'year', label: 'Рік', widget: 'number', min: 1900, max: 2200, step: 1 },
      { name: 'stack', label: 'Стек', widget: 'list', help: 'Через кому: Unity, C#' },
      { name: 'url', label: 'Посилання', widget: 'url' },
      { name: 'repo', label: 'Репозиторій', widget: 'url' },
    ],
  },
  {
    name: 'links',
    label: 'Посилання',
    singular: 'посилання',
    kind: 'yaml',
    path: 'src/data/links.yaml',
    rootKey: 'links',
    canCreate: true,
    canDelete: true,
    fields: [
      { name: 'title', label: 'Назва', widget: 'text', required: true },
      { name: 'url', label: 'Адреса', widget: 'url', required: true },
      { name: 'group', label: 'Група', widget: 'text', help: 'Блоги, Інструменти, …' },
      { name: 'note', label: 'Нотатка', widget: 'textarea' },
    ],
  },
];

export function getCollection(name: string): Collection | undefined {
  return COLLECTIONS.find((collection) => collection.name === name);
}
